go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -0,0 +1,584 @@
+			if err != nil {
+				params.Logger.Errorf("Not able to set super_read_only to its original value after restore")
+			}
+		}, nil
+	}
+
+	return func() {}, nil
+}
+
+// releaseReadLock will keep reading the MySQL Shell STDERR waiting until the point it has acquired its lock
+func releaseReadLock(ctx context.Context, reader io.Reader, params BackupParams, wg *sync.WaitGroup, lockAcquired time.Time) {
+	defer wg.Done()
+
+	scanner := bufio.NewScanner(reader)
+	released := false
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if !released {
+
+			if !strings.Contains(line, "Global read lock has been released") {
+				continue
+			}
+			released = true
+
+			params.Logger.Infof("mysql shell released its global read lock, doing the same")
+
+			err := params.Mysqld.ReleaseGlobalReadLock(ctx)
+			if err != nil {
+				params.Logger.Errorf("unable to release global read lock: %v", err)
+			}
+
+			params.Logger.Infof("global read lock released after %v", time.Since(lockAcquired))
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		params.Logger.Errorf("error reading from reader: %v", err)
+	}
+}
+
+func cleanupMySQL(ctx context.Context, params RestoreParams, shouldDeleteUsers bool) error {
+	params.Logger.Infof("Cleaning up MySQL ahead of a restore")
+	result, err := params.Mysqld.FetchSuperQuery(ctx, "SHOW DATABASES")
+	if err != nil {
+		return err
+	}
+
+	// drop all databases
+	for _, row := range result.Rows {
+		dbName := row[0].ToString()
+		if slices.Contains(internalDBs, dbName) {
+			continue // not dropping internal DBs
+		}
+
+		params.Logger.Infof("Dropping DB %q", dbName)
+		err = params.Mysqld.ExecuteSuperQuery(ctx, fmt.Sprintf("DROP DATABASE IF EXISTS `%s`", row[0].ToString()))
+		if err != nil {
+			return fmt.Errorf("error droppping database %q: %w", row[0].ToString(), err)
+		}
+	}
+
+	if shouldDeleteUsers {
+		// get current user
+		var currentUser string
+		result, err = params.Mysqld.FetchSuperQuery(ctx, "SELECT user()")
+		if err != nil {
+			return fmt.Errorf("error fetching current user: %w", err)
+		}
+
+		for _, row := range result.Rows {
+			currentUser = row[0].ToString()
+		}
+
+		// drop all users except reserved ones
+		result, err = params.Mysqld.FetchSuperQuery(ctx, "SELECT user, host FROM mysql.user")
+		if err != nil {
+			return err
+		}
+
+		for _, row := range result.Rows {
+			user := fmt.Sprintf("%s@%s", row[0].ToString(), row[1].ToString())
+
+			if user == currentUser {
+				continue // we don't drop the current user
+			}
+			if slices.Contains(reservedUsers, user) {
+				continue // we skip reserved MySQL users
+			}
+
+			params.Logger.Infof("Dropping User %q", user)
+			err = params.Mysqld.ExecuteSuperQuery(ctx, fmt.Sprintf("DROP USER '%s'@'%s'", row[0].ToString(), row[1].ToString()))
+			if err != nil {
+				return fmt.Errorf("error droppping user %q: %w", user, err)
+			}
+		}
+	}
+
+	return err
+}
+
+func init() {
+	BackupRestoreEngineMap[mysqlShellBackupEngineName] = &MySQLShellBackupEngine{}
+}
```

Comment:
We already have an `init` at the top of this file, nit picking, but we could merge them together to make it easier to read.

Reply:
ah, good catch, I have merged them together